metldata/artifacts_rest/artifact_dao.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""DAO for managing artifacts."""
+
+from hexkit.protocols.dao import DaoFactoryProtocol, DaoNaturalId
+from typing_extensions import TypeAlias
+
+from metldata.artifacts_rest.artifact_info import get_artifact_info_dict
+from metldata.artifacts_rest.models import ArtifactInfo, ArtifactResource
+
+ArtifactResourceDao: TypeAlias = DaoNaturalId[ArtifactResource]
+
+
+class DaoNotFoundError(RuntimeError):
+    """Raised when a DAO is not found."""
+
+    def __init__(self, artifact_name: str, class_name: str):
+        super().__init__(
+            f"Could not find DAO for artifact '{artifact_name}' and class '{class_name}'."
+        )
+
+
+class ArtifactDaoCollection:
+    """A collection of DAO for managing artifacts and their resources."""
+
+    @classmethod
+    async def construct(
+        cls,
+        dao_factory: DaoFactoryProtocol,
+        artifact_infos: list[ArtifactInfo],
+    ):
+        """Initialize the collection of DAOs."""
+
+        artifact_info_dict = get_artifact_info_dict(artifact_infos=artifact_infos)
+
+        # Cannot use comprehensions because of bug: https://bugs.python.org/issue33346
```

Comment:
Another reason to upgrade to Python 3.10 or 3.11 (is it already on our list?)

Reply:
No good point. I added it.